src/rtp.cpp

```diff
@@ -0,0 +1,190 @@
+    csrc_count(csrc_count_ + 1);
+}
+
+bool RTP::remove_csrc_id(const uint32_t csrc_id) {
+    small_uint<4> csrc_count_ = csrc_count();
+    if (csrc_count_ == 0) {
+        return false;
+    }
+
+    csrc_ids_type::iterator iter = search_csrc_id_iterator(Endian::host_to_be(csrc_id));
+    if (iter == csrc_ids_.end()) {
+        return false;
+    }
+
+    csrc_ids_.erase(iter);
+    csrc_count(csrc_count_ - 1);
+    return true;
+}
+
+bool RTP::search_csrc_id(const uint32_t csrc_id) {
+    csrc_ids_type::const_iterator iter = search_csrc_id_iterator(Endian::host_to_be(csrc_id));
+    return (iter != csrc_ids_.cend());
+}
+
+RTP::csrc_ids_type::const_iterator RTP::search_csrc_id_iterator(const uint32_t csrc_id) const {
+    return std::find(csrc_ids_.cbegin(), csrc_ids_.cend(), csrc_id);
+}
+
+RTP::csrc_ids_type::iterator RTP::search_csrc_id_iterator(const uint32_t csrc_id) {
+    return std::find(csrc_ids_.begin(), csrc_ids_.end(), csrc_id);
+}
+
+void RTP::add_extension_data(const uint32_t value) {
+    if (TINS_UNLIKELY(extension_length() >= 65535)) {
+        throw logic_error("Maximum number of extension data reached");
+    }
+
+    extension_bit(1);
+    ext_data_.push_back(Endian::host_to_be(value));
+    extension_length(extension_length() + 1);
+}
+
+bool RTP::remove_extension_data(const uint32_t value) {
+    if (extension_bit() == 0 || extension_length() == 0) {
+        return false;
+    }
+
+    extension_header_data_type::iterator iter = search_extension_data_iterator(Endian::host_to_be(value));
+    if (iter == ext_data_.end()) {
+        return false;
+    }
+
+    ext_data_.erase(iter);
+
+    extension_length(extension_length() - 1);
+
+    if (extension_length() == 0) {
+        extension_bit(0);
+    }
+
+    return true;
+}
+
+bool RTP::search_extension_data(const uint32_t value) {
+    if (extension_bit() == 0 || extension_length() == 0) {
+        return false;
+    }
+
+    extension_header_data_type::const_iterator iter = search_extension_data_iterator(Endian::host_to_be(value));
+    return (iter != ext_data_.cend());
+}
+
+RTP::extension_header_data_type::const_iterator RTP::search_extension_data_iterator(const uint32_t data) const {
+    return std::find(ext_data_.cbegin(), ext_data_.cend(), data);
+}
+
+RTP::extension_header_data_type::iterator RTP::search_extension_data_iterator(const uint32_t data) {
+    return std::find(ext_data_.begin(), ext_data_.end(), data);
+}
+
+void RTP::write_serialization(uint8_t* buffer, uint32_t total_sz) {
+    OutputMemoryStream stream(buffer, total_sz);
+    stream.write(header_);
+
+    for (auto csrc_id : csrc_ids_) {
+        stream.write(csrc_id);
+    }
+
+    if (extension_bit() == 1) {
+        stream.write(ext_header_);
+        for (auto data : ext_data_) {
+            stream.write(data);
+        }
+    }
+
+    if (padding_bit() == 1) {
+        if (inner_pdu()) {
+            stream.skip(inner_pdu()->size());
+        }
+        stream.fill(padding_size_ - 1, 0);
```

Comment:
If you call `padding_bit(1)` but you don't set a padding size, this underflows and blows up. I think it may make sense to only support `padding_size(...)` and let that set the bit like it's currently doing. Allowing to set the bit manually can lead to these types of errors.

Reply:
That makes sense.

I have made the `padding_bit` setter function hidden from the public interface of the RTP class. I have also modified `write_serialization` to throw a `pdu_not_serializable` exception if the padding bit is set but the padding size is still zero.